GHC 9.4 compatibility

If I override the bounds, I currently get this error, building with GHC 9.4:

```
Building library for ghc-source-gen-0.4.3.0..
[ 1 of 18] Compiling GHC.SourceGen.Pretty ( src/GHC/SourceGen/Pretty.hs, dist/build/GHC/SourceGen/Pretty.o, dist/build/GHC/SourceGen/Pretty.dyn_o )
[ 2 of 18] Compiling GHC.SourceGen.Syntax.Internal ( src/GHC/SourceGen/Syntax/Internal.hs, dist/build/GHC/SourceGen/Syntax/Internal.o, dist/build/GHC/SourceGen/Syntax/Internal.dyn_o )
[ 3 of 18] Compiling GHC.SourceGen.Name.Internal ( src/GHC/SourceGen/Name/Internal.hs, dist/build/GHC/SourceGen/Name/Internal.o, dist/build/GHC/SourceGen/Name/Internal.dyn_o )
[ 4 of 18] Compiling GHC.SourceGen.Name ( src/GHC/SourceGen/Name.hs, dist/build/GHC/SourceGen/Name.o, dist/build/GHC/SourceGen/Name.dyn_o )
[ 5 of 18] Compiling GHC.SourceGen.Lit.Internal ( src/GHC/SourceGen/Lit/Internal.hs, dist/build/GHC/SourceGen/Lit/Internal.o, dist/build/GHC/SourceGen/Lit/Internal.dyn_o )
[ 6 of 18] Compiling GHC.SourceGen.Pat.Internal ( src/GHC/SourceGen/Pat/Internal.hs, dist/build/GHC/SourceGen/Pat/Internal.o, dist/build/GHC/SourceGen/Pat/Internal.dyn_o )

src/GHC/SourceGen/Pat/Internal.hs:54:27: error:
    • Couldn't match type: Language.Haskell.Syntax.Extension.XRec
                             p0 (Pat p0)
                           -> Language.Haskell.Syntax.Extension.XRec
                                p0 (Language.Haskell.Syntax.Extension.HsToken ")")
                           -> Pat p0
                     with: Pat GHC.Hs.Extension.GhcPs
      Expected: GHC.Parser.Annotation.EpAnn ann0
                -> Language.Haskell.Syntax.Extension.XRec
                     p0 (Language.Haskell.Syntax.Extension.HsToken "(")
                -> Pat'
        Actual: Language.Haskell.Syntax.Extension.XParPat p0
                -> Language.Haskell.Syntax.Extension.LHsToken "(" p0
                -> Language.Haskell.Syntax.Pat.LPat p0
                -> Language.Haskell.Syntax.Extension.LHsToken ")" p0
                -> Pat p0
    • Probable cause: ‘ParPat’ is applied to too few arguments
      In the first argument of ‘withEpAnnNotUsed’, namely ‘ParPat’
      In the first argument of ‘(.)’, namely ‘withEpAnnNotUsed ParPat’
      In the expression: withEpAnnNotUsed ParPat . builtPat
   |
54 | parPat = withEpAnnNotUsed ParPat . builtPat
   |                           ^^^^^^
```

--- a/ghc-source-gen.cabal
+++ b/ghc-source-gen.cabal
@@ -104,7 +104,7 @@ test-suite name_test
   build-depends:
       QuickCheck >=2.10 && <2.15
     , base >=4.7 && <5
-    , ghc >=8.4 && <9.3
+    , ghc >=8.4 && <9.5
     , ghc-source-gen
     , tasty >=1.0 && <1.5
     , tasty-hunit ==0.10.*
@@ -125,7 +125,7 @@ test-suite pprint_examples
       TypeSynonymInstances
   build-depends:
       base >=4.7 && <5
-    , ghc >=8.4 && <9.3
+    , ghc >=8.4 && <9.5
     , ghc-paths ==0.1.*
     , ghc-source-gen
     , tasty >=1.0 && <1.5
@@ -151,7 +151,7 @@ test-suite pprint_test
       TypeSynonymInstances
   build-depends:
       base >=4.7 && <5
-    , ghc >=8.4 && <9.3
+    , ghc >=8.4 && <9.5
     , ghc-paths ==0.1.*
     , ghc-source-gen
     , tasty >=1.0 && <1.5

--- a/ghc-source-gen.cabal
+++ b/ghc-source-gen.cabal
@@ -61,7 +61,7 @@ library
       TypeSynonymInstances
   build-depends:
       base >=4.7 && <5
-    , ghc >=8.4 && <9.3
+    , ghc >=8.4 && <9.5
   if impl(ghc<8.10)
     other-modules:
         GHC.Hs

--- a/package.yaml
+++ b/package.yaml
@@ -30,7 +30,7 @@ description: |
 
 dependencies:
 - base >= 4.7 && < 5
-- ghc >= 8.4 && < 9.3
+- ghc >= 8.4 && < 9.5
 
 default-extensions:
 - DataKinds

--- a/src/GHC/SourceGen/Binds/Internal.hs
+++ b/src/GHC/SourceGen/Binds/Internal.hs
@@ -96,7 +96,11 @@ matchGroup context matches =
 
 mkGRHSs :: RawGRHSs -> GRHSs' LHsExpr'
 mkGRHSs g = withEmptyEpAnnComments GRHSs
+#if MIN_VERSION_ghc(9,4,0)
+                (map mkLocated $ rawGRHSs g)
+#else
                 (map builtLoc $ rawGRHSs g)
+#endif
                 (fromLocalBinds $ valBinds $ rawGRHSWhere g)
   where
 #if MIN_VERSION_ghc(9,2,0)

--- a/src/GHC/SourceGen/Decl.hs
+++ b/src/GHC/SourceGen/Decl.hs
@@ -328,7 +328,11 @@ newOrDataType newOrData name vars conDecls derivs
                 Nothing
                 Nothing
                 (map mkLocated conDecls)
+#if MIN_VERSION_ghc(9,4,0)
+                (toHsDeriving $ map mkLocated derivs)
+#else
                 (toHsDeriving $ map builtLoc derivs)
+#endif
   where
 #if MIN_VERSION_ghc(9,2,0)
     cxt = Nothing
@@ -402,7 +406,11 @@ recordCon name fields = renderCon98Decl name
   where
     mkLConDeclField (n, f) =
         mkLocated $ withEpAnnNotUsed ConDeclField
+#if MIN_VERSION_ghc(9,4,0)
+                        [mkLocated $ withPlaceHolder $ noExt FieldOcc $ valueRdrName $ unqual n]
+#else
                         [builtLoc $ withPlaceHolder $ noExt FieldOcc $ valueRdrName $ unqual n]
+#endif
                         (renderField f)
                         Nothing
 
@@ -471,7 +479,11 @@ deriving' = derivingWay Nothing
 
 derivingWay :: Maybe DerivStrategy' -> [HsType'] -> HsDerivingClause'
 derivingWay way ts =
+#if MIN_VERSION_ghc(9,4,0)
+    withEpAnnNotUsed HsDerivingClause (fmap mkLocated way) $ mkLocated $ derivClauseTys $ map sigType ts
+#else
     withEpAnnNotUsed HsDerivingClause (fmap builtLoc way) $ mkLocated $ derivClauseTys $ map sigType ts
+#endif
   where
 #if MIN_VERSION_ghc(9,2,0)
     derivClauseTys [x] = noExt DctSingle x
@@ -557,7 +569,11 @@ standaloneDerivingAnyclass = standaloneDerivingWay (Just strat)
 standaloneDerivingWay :: Maybe DerivStrategy' -> HsType' -> HsDecl'
 standaloneDerivingWay way ty = noExt DerivD derivDecl
   where derivDecl =
+#if MIN_VERSION_ghc(9,4,0)
+          withEpAnnNotUsed DerivDecl (hsWC $ sigType ty) (fmap mkLocated way) Nothing
+#else
           withEpAnnNotUsed DerivDecl (hsWC $ sigType ty) (fmap builtLoc way) Nothing
+#endif
         hsWC =
 #if MIN_VERSION_ghc(8,6,0)
           noExt HsWC

--- a/src/GHC/SourceGen/Expr.hs
+++ b/src/GHC/SourceGen/Expr.hs
@@ -30,7 +30,11 @@ module GHC.SourceGen.Expr
 
 import GHC.Hs.Expr
 import GHC.Hs.Extension (GhcPs)
+#if MIN_VERSION_ghc(9,4,0)
+import GHC.Hs.Pat (HsFieldBind(..), HsRecFields(..))
+#else
 import GHC.Hs.Pat (HsRecField'(..), HsRecFields(..))
+#endif
 import GHC.Hs.Type (FieldOcc(..), AmbiguousFieldOcc(..))
 import GHC.Hs.Utils (mkHsIf)
 import Data.String (fromString)
@@ -70,7 +74,11 @@ overLabel = hsOverLabel . fromString
 #endif
 
 let' :: [RawValBind] -> HsExpr' -> HsExpr'
+#if MIN_VERSION_ghc(9,4,0)
+let' binds e = withEpAnnNotUsed HsLet mkToken (toHsLocalBinds $ valBinds binds) mkToken $ mkLocated e
+#else
 let' binds e = withEpAnnNotUsed HsLet (toHsLocalBinds $ valBinds binds) $ mkLocated e
+#endif
   where
 #if MIN_VERSION_ghc(9,2,0)
     toHsLocalBinds = id
@@ -86,7 +94,11 @@ lambda :: [Pat'] -> HsExpr' -> HsExpr'
 lambda ps e = noExt HsLam $ matchGroup LambdaExpr [match ps e]
 
 lambdaCase :: [RawMatch] -> HsExpr'
+#if MIN_VERSION_ghc(9,4,0)
+lambdaCase = withEpAnnNotUsed HsLamCase LamCase . matchGroup CaseAlt
+#else
 lambdaCase = withEpAnnNotUsed HsLamCase . matchGroup CaseAlt
+#endif
 
 if' :: HsExpr' -> HsExpr' -> HsExpr' -> HsExpr'
 if' x y z = mkHsIf
@@ -109,7 +121,11 @@ if' x y z = mkHsIf
 -- >     , guardedStmt (var "otherwise") $ rhs (string "h")
 -- >     ]
 multiIf :: [GuardedExpr] -> HsExpr'
+#if MIN_VERSION_ghc(9,4,0)
+multiIf = withPlaceHolder (withEpAnnNotUsed HsMultiIf) . map mkLocated
+#else
 multiIf = withPlaceHolder (withEpAnnNotUsed HsMultiIf) . map builtLoc
+#endif
 
 -- | A do-expression.
 --
@@ -210,6 +226,14 @@ recordConE c fs = (withPlaceHolder $ withEpAnnNotUsed RecordCon (valueRdrName c)
   where
     recField :: (RdrNameStr, HsExpr') -> LHsRecField' LHsExpr'
     recField (f, e) =
+#if MIN_VERSION_ghc(9,4,0)
+        mkLocated HsFieldBind
+            { hfbLHS =
+                  mkLocated $ withPlaceHolder $ noExt FieldOcc $ valueRdrName f
+            , hfbRHS = mkLocated e
+            , hfbPun = False
+            , hfbAnn = EpAnnNotUsed
+#else
         mkLocated HsRecField
             { hsRecFieldLbl =
                   builtLoc $ withPlaceHolder $ noExt FieldOcc $ valueRdrName f
@@ -217,6 +241,7 @@ recordConE c fs = (withPlaceHolder $ withEpAnnNotUsed RecordCon (valueRdrName c)
             , hsRecPun = False
 #if MIN_VERSION_ghc(9,2,0)
             , hsRecFieldAnn = EpAnnNotUsed
+#endif
 #endif
             }
 
@@ -241,6 +266,14 @@ recordUpd e fs =
   where
     mkField :: (RdrNameStr, HsExpr') -> LHsRecUpdField'
     mkField (f, e') =
+#if MIN_VERSION_ghc(9,4,0)
+        mkLocated HsFieldBind
+            { hfbLHS =
+                mkLocated $ withPlaceHolder $ noExt Ambiguous $ valueRdrName f
+            , hfbRHS = mkLocated e'
+            , hfbPun = False
+            , hfbAnn = EpAnnNotUsed
+#else
         mkLocated HsRecField
             { hsRecFieldLbl =
                 builtLoc $ withPlaceHolder $ noExt Ambiguous $ valueRdrName f
@@ -248,6 +281,7 @@ recordUpd e fs =
             , hsRecPun = False
 #if MIN_VERSION_ghc(9,2,0)
             , hsRecFieldAnn = EpAnnNotUsed
+#endif
 #endif
             }
     withPlaceHolder4 = withPlaceHolder . withPlaceHolder . withPlaceHolder

--- a/src/GHC/SourceGen/Expr/Internal.hs
+++ b/src/GHC/SourceGen/Expr/Internal.hs
@@ -17,6 +17,10 @@ import SrcLoc (unLoc)
 import GHC.SourceGen.Lit.Internal
 import GHC.SourceGen.Syntax.Internal
 
+#if MIN_VERSION_ghc(9,4,0)
+import Language.Haskell.Syntax.Extension
+#endif
+
 parenthesizeExprForApp, parenthesizeExprForOp
     :: LHsExpr' -> LHsExpr'
 parenthesizeExprForApp e 
@@ -27,7 +31,11 @@ parenthesizeExprForOp e
     | otherwise = e
 
 parExpr :: LHsExpr' -> LHsExpr'
+#if MIN_VERSION_ghc(9,4,0)
+parExpr e = mkLocated $ withEpAnnNotUsed HsPar mkToken e mkToken
+#else
 parExpr = mkLocated . withEpAnnNotUsed HsPar
+#endif
 
 #if MIN_VERSION_ghc(8,6,0)
 #define WILD_EXT _

--- a/src/GHC/SourceGen/Lit.hs
+++ b/src/GHC/SourceGen/Lit.hs
@@ -45,7 +45,11 @@ instance HasLit HsExpr' where
 instance HasLit Pat' where
     lit = noExt LitPat
     overLit l = withPlaceHolder
+#if MIN_VERSION_ghc(9,4,0)
+                    $ withEpAnnNotUsed NPat (mkLocated l) Nothing noSyntaxExpr
+#else
                     $ withEpAnnNotUsed NPat (builtLoc l) Nothing noSyntaxExpr
+#endif
 
 char :: HasLit e => Char -> e
 char = lit . noSourceText HsChar
@@ -55,13 +59,21 @@ string = lit . noSourceText HsString . fsLit
 
 -- | Note: this is an *overloaded* integer.
 int :: HasLit e => Integer -> e
+#if MIN_VERSION_ghc(9,4,0)
+int n = overLit $ withPlaceHolder $ withPlaceHolder (noExt OverLit n')
+#else
 int n = overLit $ withPlaceHolder $ withPlaceHolder (noExt OverLit n') noExpr
+#endif
   where
     n' = HsIntegral $ mkIntegralLit n
 
 -- | Note: this is an *overloaded* rational, e.g., a decimal number.
 frac :: HasLit e => Rational -> e
+#if MIN_VERSION_ghc(9,4,0)
+frac x = overLit $ withPlaceHolder $ withPlaceHolder (noExt OverLit $ HsFractional x')
+#else
 frac x = overLit $ withPlaceHolder $ withPlaceHolder (noExt OverLit $ HsFractional x') noExpr
+#endif
   where
 #if MIN_VERSION_ghc(9,2,0)
     x' = mkTHFractionalLit x

--- a/src/GHC/SourceGen/Module.hs
+++ b/src/GHC/SourceGen/Module.hs
@@ -44,6 +44,9 @@ import GHC.Types.Name.Reader (RdrName)
 #else
 import RdrName (RdrName)
 #endif
+#if MIN_VERSION_ghc(9,4,0)
+import GHC.Types.PkgQual (RawPkgQual(..))
+#endif
 
 import GHC.SourceGen.Syntax.Internal
 import GHC.SourceGen.Name
@@ -86,7 +89,11 @@ as' d m = d { ideclAs = Just (mkLocated $ unModuleNameStr m) }
 import' :: ModuleNameStr -> ImportDecl'
 import' m = noSourceText (withEpAnnNotUsed ImportDecl)
             (mkLocated $ unModuleNameStr m)
-            Nothing
+#if MIN_VERSION_ghc(9,4,0)
+            NoRawPkgQual
+#else
+            Nothing 
+#endif
 #if MIN_VERSION_ghc(9,0,0)
             NotBoot
 #else

--- a/src/GHC/SourceGen/Overloaded.hs
+++ b/src/GHC/SourceGen/Overloaded.hs
@@ -56,10 +56,18 @@ class Par e where
     par :: e -> e
 
 instance Par HsExpr' where
+#if MIN_VERSION_ghc(9,4,0)
+    par p = withEpAnnNotUsed HsPar mkToken (mkLocated p) mkToken
+#else
     par = withEpAnnNotUsed HsPar . mkLocated
+#endif
 
 instance Par Pat' where
+#if MIN_VERSION_ghc(9,4,0)
+    par p = withEpAnnNotUsed ParPat mkToken (builtPat p) mkToken
+#else
     par = withEpAnnNotUsed ParPat . builtPat
+#endif
 
 instance Par HsType' where
     par = withEpAnnNotUsed HsParTy . mkLocated
@@ -139,9 +147,15 @@ instance App HsExpr' where
 
 instance App HsType' where
     op x o y
+#if MIN_VERSION_ghc(9,4,0)
+        = withEpAnnNotUsed HsOpTy notPromoted (parenthesizeTypeForOp $ mkLocated x)
+                (typeRdrName o)
+                (parenthesizeTypeForOp $ mkLocated y)
+#else
         = noExt HsOpTy (parenthesizeTypeForOp $ mkLocated x)
                 (typeRdrName o)
                 (parenthesizeTypeForOp $ mkLocated y)
+#endif
     x @@ y = noExt HsAppTy
                 (parenthesizeTypeForOp $ mkLocated x)
                 (parenthesizeTypeForApp $ mkLocated y)

--- a/src/GHC/SourceGen/Pat.hs
+++ b/src/GHC/SourceGen/Pat.hs
@@ -81,6 +81,13 @@ recordConP c fs =
   where
     mkRecField :: (RdrNameStr, Pat') -> LHsRecField' LPat'
     mkRecField (f, p) =
+#if MIN_VERSION_ghc(9,4,0)
+        mkLocated $ HsFieldBind
+            { hfbAnn = EpAnnNotUsed
+            , hfbLHS = mkLocated $ withPlaceHolder $ noExt FieldOcc $ valueRdrName f
+            , hfbRHS = builtPat p
+            , hfbPun = False
+#else
         mkLocated $ HsRecField
             { hsRecFieldLbl =
                 builtLoc $ withPlaceHolder $ noExt FieldOcc $ valueRdrName f
@@ -88,6 +95,7 @@ recordConP c fs =
             , hsRecPun = False
 #if MIN_VERSION_ghc(9,2,0)
             , hsRecFieldAnn = EpAnnNotUsed
+#endif
 #endif
             }
 

--- a/src/GHC/SourceGen/Pat/Internal.hs
+++ b/src/GHC/SourceGen/Pat/Internal.hs
@@ -51,5 +51,9 @@ needsPar SigPatOut{} = True
 needsPar _ = False
 
 parPat :: Pat' -> Pat'
+#if MIN_VERSION_ghc(9,4,0)
+parPat p = withEpAnnNotUsed ParPat mkToken (builtPat p) mkToken
+#else
 parPat = withEpAnnNotUsed ParPat . builtPat
+#endif
 

--- a/src/GHC/SourceGen/Syntax/Internal.hs
+++ b/src/GHC/SourceGen/Syntax/Internal.hs
@@ -93,6 +93,10 @@ import qualified PlaceHolder as GHC
 import GHC.Types.Var (Specificity)
 #endif
 
+#if MIN_VERSION_ghc(9,4,0)
+import GHC.Parser.Annotation
+#endif
+
 import GHC.Hs.Extension (GhcPs)
 
 #if MIN_VERSION_ghc(8,10,0)
@@ -339,4 +343,14 @@ type HsPatSigType' = LHsSigWcType'
 type LIdP = GHC.LIdP GHC.GhcPs
 #else
 type LIdP = Located (GHC.IdP GHC.GhcPs)
+#endif
+
+#if MIN_VERSION_ghc(9,4,0)
+mkUniToken :: GenLocated TokenLocation (GHC.HsUniToken t u)
+mkUniToken = L NoTokenLoc GHC.HsNormalTok
+#endif
+
+#if MIN_VERSION_ghc(9,4,0)
+mkToken :: GenLocated TokenLocation (GHC.HsToken t)
+mkToken = L NoTokenLoc GHC.HsTok
 #endif

--- a/src/GHC/SourceGen/Type.hs
+++ b/src/GHC/SourceGen/Type.hs
@@ -29,6 +29,10 @@ import GHC.Parser.Annotation
 import GHC.Hs.Type
 #endif
 
+#if MIN_VERSION_ghc(9,4,0)
+import Language.Haskell.Syntax.Extension
+#endif
+
 import GHC.SourceGen.Syntax.Internal
 import GHC.SourceGen.Lit.Internal (noSourceText)
 import GHC.SourceGen.Name.Internal
@@ -62,7 +66,9 @@ tuplePromotedTy = withPlaceHolders (withEpAnnNotUsed HsExplicitTupleTy) . map mk
 -- > var "a" --> var "b"
 (-->) :: HsType' -> HsType' -> HsType'
 a --> b = withEpAnnNotUsed HsFunTy
-#if MIN_VERSION_ghc(9,0,0)
+#if MIN_VERSION_ghc(9,4,0)
+         (HsUnrestrictedArrow mkUniToken)
+#elif MIN_VERSION_ghc(9,0,0)
          (HsUnrestrictedArrow NormalSyntax)
 #endif
          (parenthesizeTypeForFun $ mkLocated a) (mkLocated b)
@@ -96,7 +102,9 @@ forall' ts = noExt hsForAllTy (map mkLocated ts) . mkLocated
 (==>) :: [HsType'] -> HsType' -> HsType'
 (==>) cs = hsQualTy (mkLocated (map mkLocated cs)) . mkLocated
   where
-#if MIN_VERSION_ghc(9,2,0)
+#if MIN_VERSION_ghc(9,4,0)
+    hsQualTy = noExt HsQualTy 
+#elif MIN_VERSION_ghc(9,2,0)
     hsQualTy = noExt HsQualTy . Just
 #else
     hsQualTy = noExt HsQualTy

